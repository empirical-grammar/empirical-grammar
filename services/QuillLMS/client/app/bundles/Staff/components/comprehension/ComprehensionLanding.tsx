import * as React from "react";
import { NavLink, Redirect, Route, RouteComponentProps, Switch, withRouter } from 'react-router-dom';
import { queryCache } from 'react-query';

import ActivityForm from './configureSettings/activityForm';
import Activities from './activities';
import Activity from './activity';
import SubmissionModal from './shared/submissionModal';

import { blankActivity } from '../../../../constants/comprehension';
import { ActivityInterface } from '../../interfaces/comprehensionInterfaces';
import { createActivity } from '../../utils/comprehension/activityAPIs';
import { getCsrfToken } from "../../helpers/comprehension";
import { Modal } from '../../../Shared/index';

const ComprehensionLanding = ({ location }: RouteComponentProps) => {
  const { pathname } = location
  const [showCreateActivityModal, setShowCreateActivityModal] = React.useState<boolean>(false);
  const [showSubmissionModal, setShowSubmissionModal] = React.useState<boolean>(false);
  const [error, setError] = React.useState<string>('');

  const csrfToken = getCsrfToken();
  localStorage.setItem('csrfToken', csrfToken);

  const checkIndexActive = () => {
    if(!location) return false;
    return pathname === '/activities' && !showCreateActivityModal;
  }

  const checkOverviewActive = () => {
    if(!location) return false;
    return pathname !== '/activities' && !showCreateActivityModal;
  }

  const toggleCreateActivityModal = () => {
    setShowCreateActivityModal(!showCreateActivityModal);
  }

  const toggleSubmissionModal = () => {
    setShowSubmissionModal(!showSubmissionModal);
  }

  const submitActivity = (activity: ActivityInterface) => {
    createActivity(activity).then((response) => {
      const { error } = response;
      if(error) setError(error);
      setShowCreateActivityModal(false);
      setShowSubmissionModal(true);

      // update activities cache to display newly created activity
      queryCache.refetchQueries('activities')
    });
  }

  const renderActivityForm = () => {
    return(
      <Modal>
        <ActivityForm activity={blankActivity} closeModal={toggleCreateActivityModal} submitActivity={submitActivity} />
      </Modal>
    );
  }

  const renderSubmissionModal = () => {
    const message = error ? error : 'Submission successful!';
    return <SubmissionModal close={toggleSubmissionModal} message={message} />;
  }

  return(
    <div className="main-admin-container">
      <section className="left-side-menu">
        <p className="menu-label">
          Home
        </p>
        <ul className="menu-list">
          <NavLink activeClassName='is-active' isActive={checkIndexActive} to='/activities'>
            Activities Index
          </NavLink>
          <NavLink activeClassName='is-active' isActive={checkOverviewActive} to={pathname}>
            Activity Overview
          </NavLink>
          <button className={`create-activity-button ${showCreateActivityModal ? 'is-active' :''}`} onClick={toggleCreateActivityModal} type="submit">
            Create New Activity
          </button>
        </ul>
      </section>
      {showCreateActivityModal && renderActivityForm()}
      {showSubmissionModal && renderSubmissionModal()}
      <Switch>
        <Redirect exact from='/' to='/activities' />
        <Route component={Activity} path='/activities/:activityId' />
        <Route component={Activities} path='/activities' />
      </Switch>
    </div>
  )
}

export default withRouter(ComprehensionLanding);
