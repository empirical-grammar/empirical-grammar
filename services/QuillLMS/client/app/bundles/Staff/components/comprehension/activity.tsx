import * as React from "react";
import { Redirect, Route, RouteComponentProps, Switch, withRouter } from 'react-router-dom';

import Navigation from './navigation'
import ActivitySettings from './configureSettings/activitySettings';
import Rules from './configureRules/rules';
import Rule from './configureRules/rule';
import RulesAnalysis from './rulesAnalysis/rulesAnalysis';
import RuleAnalysis from './rulesAnalysis/ruleAnalysis';
import TurkSessions from './gatherResponses/turkSessions';
import SemanticLabelsIndex from './semanticRules/semanticLabelsIndex';

import { ActivityRouteProps } from '../../interfaces/comprehensionInterfaces';

const Activity: React.FC<RouteComponentProps<ActivityRouteProps>> = ({ match, location, }) => {
  const { params } = match;
  return(<React.Fragment>
    <Navigation location={location} match={match} />
    <div className="activity-container">
      <Switch>
        <Redirect exact from='/activities/:activityId' to='/activities/:activityId/settings' />
        <Route component={ActivitySettings} path='/activities/:activityId/settings' />
        <Route component={Rule} path='/activities/:activityId/rules/:ruleId' />
        <Route component={Rules} path='/activities/:activityId/rules' />
        <Route component={RuleAnalysis} path='/activities/:activityId/rules-analysis/:ruleId' />
        <Route component={RulesAnalysis} path='/activities/:activityId/rules-analysis' />
        <Route component={TurkSessions} path='/activities/:activityId/turk-sessions' />
        <Route component={SemanticLabelsIndex} path='/activities/:activityId/semantic-labels' />
      </Switch>
    </div>
  </React.Fragment>);
}
export default withRouter(Activity)
