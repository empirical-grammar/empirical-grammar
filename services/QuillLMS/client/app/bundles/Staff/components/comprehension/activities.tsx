import * as React from "react";
import { Link } from 'react-router-dom'
import { DataTable, Error, Spinner } from 'quill-component-library/dist/componentLibrary';
import { ActivityInterface } from '../../interfaces/comprehensionInterfaces';
import { blankActivity } from '../../../../constants/comprehension';
import { activitiesGetAPI } from '../../utils/comprehensionAPIs';
import useSWR from 'swr';

const Activities = () => {
  const [activities, setActivities] = React.useState<ActivityInterface[]>([blankActivity]);
  const [loading, setLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<string>(null);
  
  const fetchData = async () => {
    let activities: ActivityInterface[];
    try {
      setLoading(true);
      const response = await fetch(activitiesGetAPI);
      activities = await response.json();
    } catch (error) {
      setError(error);
      setLoading(false);
    }
    setActivities(activities);
    setLoading(false);
  };

  // cache activity data for updates
  useSWR("activities", fetchData);

  React.useEffect(() => {
    fetchData();
  }, []);

  const formattedRows = activities.map((activity: ActivityInterface) => {
    const { id,  title } = activity;
    const activityLink = (<Link to={`/activities/${id}`}>{title}</Link>);
    return {
      title: activityLink
    }
  });

  if(loading) {
    return(
      <div className="loading-spinner-container">
        <Spinner />
      </div>
    );
  }

  if(error) {
    return(
      <div className="error-container">
        <Error error={`${error}`} />
      </div>
    );
  }

  const dataTableFields = [
    { name: "Title", attribute:"title", width: "900px" }
  ];

  return(
    <div className="activities-container">
      <DataTable
        className="activities-table"
        defaultSortAttribute="title"
        headers={dataTableFields}
        rows={formattedRows}
      />
    </div>
  );
}

export default Activities
