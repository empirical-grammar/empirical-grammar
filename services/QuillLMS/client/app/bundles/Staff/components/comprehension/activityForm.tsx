import * as React from "react";
import { DropdownInput, Input, TextEditor } from 'quill-component-library/dist/componentLibrary';
import { EditorState, ContentState } from 'draft-js'
import { flagOptions } from '../../../../constants/comprehension'

// TODO: re-enable form inputs for course, target reading level and reading level score

const ActivityForm = (props) => {

  const { activity, closeModal, submitActivity } = props;
  const { flag, passages, prompts } = activity;
  // const formattedCourse = course ? { label: course, value: course } : courseOptions[0];
  const formattedFlag = flag ? { label: flag, value: flag } : flagOptions[0];
  const formattedPassage = passages ? passages[0] : '';
  const becausePrompt = prompts ? prompts[0].text : '';
  const butPrompt = prompts ? prompts[1].text : '';
  const soPrompt = prompts ? prompts[2].text : '';

  const [activityTitle, setActivityTitle] = React.useState(activity.title || '');
  // const [activityCourse, setActivityCourse] = React.useState(formattedCourse);
  const [activityFlag, setActivityFlag] = React.useState(formattedFlag);
  const [activityPassage, setActivityPassage] = React.useState(formattedPassage);
  // const [activityReadingLevel, setActivityReadingLevel] = React.useState(readingLevelOptions[0]);
  // const [activityReadingScore, setActivityReadingScore] = React.useState('');
  const [activityBecausePrompt, setActivityBecausePrompt] = React.useState(becausePrompt);
  const [activityButPrompt, setActivityButPrompt] = React.useState(butPrompt);
  const [activitySoPrompt, setActivitySoPrompt] = React.useState(soPrompt);
  const [errors, setErrors] = React.useState([]);

  const handleSetActivityTitle = (e) => { setActivityTitle(e.target.value) };
  // const handleSetActivityCourse = (course) => { setActivityCourse(course) };
  const handleSetActivityFlag = (flag) => { setActivityFlag(flag) };
  const handleSetActivityPassage = (text) => { setActivityPassage(text) };
  // const handleSetActivityReadingLevel = (level) => { setActivityReadingLevel(level) };
  // const handleSetActivityReadingScore = (e) => { setActivityReadingScore(e.target.value) };
  const handleSetActivityBecausePrompt = (e) => { setActivityBecausePrompt(e.target.value) };
  const handleSetActivityButPrompt = (e) => { setActivityButPrompt(e.target.value) };
  const handleSetActivitySoPrompt = (e) => { setActivitySoPrompt(e.target.value) };

  const buildActivity = () => {
    return {
      title: activityTitle,
      flag: activityFlag,
      passages: [activityPassage],
      prompts: [becausePrompt, butPrompt, soPrompt]
    };
  }

  const validateForm = () => {
    let errors = {};
    const keys = ['Title', 'Passage', 'Because stem', 'But stem', 'So stem'];
    const state = [activityTitle, activityPassage, activityBecausePrompt, activityButPrompt, activitySoPrompt];
    state.map((value, i) => {
      // Text Editor is empty
      if(i === 1 && value === '<br/>') {
        errors[keys[i]] = `${keys[i]} cannot be blank.`;
      } else if(!value || value.length === 0) {
        errors[keys[i]] = `${keys[i]} cannot be blank.`;
      }
    });
    return errors;
  }

  const handleSubmitActivity = () => {
    // TODO: add validations for each input
    const activity = buildActivity();
    const validationErrors = validateForm();
    if(validationErrors && Object.keys(validationErrors).length !== 0) {
      setErrors(validationErrors);
    } else {
      submitActivity(activity);
    }
  }

  const errorsPresent = Object.keys(errors).length !== 0;

  return(
    <div className="activity-form-container">
      <div className="close-button-container">
        <button className="quill-button fun primary contained close-button" onClick={closeModal} type="submit">x</button>
      </div>
      <form className="activity-form">
        <Input
          className="title-input"
          error={errors['Title']}
          handleChange={handleSetActivityTitle}
          label="Title"
          value={activityTitle}
        />
        {/* <DropdownInput
          className="course-input"
          handleChange={handleSetActivityCourse}
          label="Course"
          options={courseOptions}
          value={activityCourse}
        /> */}
        <DropdownInput
          className="flag-input"
          handleChange={handleSetActivityFlag}
          label="Development Stage"
          options={flagOptions}
          value={activityFlag}
        />
        <TextEditor
          ContentState={ContentState}
          EditorState={EditorState}
          handleTextChange={handleSetActivityPassage}
          key="passage-description"
          text={activityPassage}
        />
        {errors['Passage'] && <p className="error-message">{errors['Passage']}</p>}
        {/* <DropdownInput
          className="reading-level-input"
          handleChange={handleSetActivityReadingLevel}
          label="Target Reading Level"
          options={readingLevelOptions}
          value={activityReadingLevel}
        />
        <Input
          className="reading-score-input"
          handleChange={handleSetActivityReadingScore}
          label="Reading Score Level"
          value={activityReadingScore}
        /> */}
        <Input
          className="because-input"
          error={errors['Because stem']}
          handleChange={handleSetActivityBecausePrompt}
          label="Because Stem"
          value={activityBecausePrompt}
        />
        <Input
          className="but-input"
          error={errors['But stem']}
          handleChange={handleSetActivityButPrompt}
          label="But Stem"
          value={activityButPrompt}
        />
        <Input
          className="so-input"
          error={errors['So stem']}
          handleChange={handleSetActivitySoPrompt}
          label="So Stem" 
          value={activitySoPrompt}
        />
        <div className="submit-button-container">
          {errorsPresent && <div className="error-message-container">
            <p className="all-errors-message">Please check that all fields have been completed correctly.</p>
          </div>}
          <button className="quill-button fun primary contained submit-button" onClick={handleSubmitActivity} type="submit">
            Submit
          </button>
          <button className="quill-button fun primary contained cancel-button" onClick={closeModal} type="submit">
            Cancel
          </button>
        </div>
      </form>
    </div>
  )
}

export default ActivityForm;
