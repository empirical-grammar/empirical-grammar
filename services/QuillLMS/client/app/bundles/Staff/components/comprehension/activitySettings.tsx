import * as React from "react";
import { DataTable, DropdownInput, Error, Modal, Spinner } from 'quill-component-library/dist/componentLibrary';
import { ActivityInterface } from '../../interfaces/comprehension/activityInterface'
import ActivityForm from './activityForm'
import { flagOptions } from '../../../../constants/comprehension'

const ActivitySettings = (props: any) => {
  const [activity, setActivity] = React.useState<ActivityInterface>({});
  const [loading, setLoading] = React.useState(false);
  const [error, setError] = React.useState(null);
  const [flag, setFlag] = React.useState(null);
  const [showEditActivityModal, setShowEditActivityModal] = React.useState(false)
  const [showEditFlagModal, setShowEditFlagModal] = React.useState(false)
  const { match } = props;
  const { params } = match;
  const { activityId } = params;
  const fetchActivityAPI = `https://comprehension-dummy-data.s3.us-east-2.amazonaws.com/activities/${activityId}.json`
  
  const fetchData = async () => {
    try {
      setLoading(true);
      const response = await fetch(fetchActivityAPI);
      var activity = await response.json();
    } catch (error) {
      setError(error);
      setLoading(false);
    }
    const { flag } = activity
    setActivity(activity);
    setFlag({ label: flag, value: flag });
    setLoading(false);
  };

  React.useEffect(() => {
    fetchData();
  }, []);

  const submitActivity = (activity) => {
    // TODO: hook into Activity PUT API
    toggleEditActivityModal();
  }

  const handleUpdateFlag = () => {
    // TODO: hook into Activity PUT API for updating only the development status (as requested by curriculum)
    toggleFlagModal();
  }

  const handleFlagChange = (flag) => {
    setFlag(flag);
  }


  const toggleEditActivityModal = () => {
    setShowEditActivityModal(!showEditActivityModal);
  }

  const toggleFlagModal = () => {
    setShowEditFlagModal(!showEditFlagModal);
  }

  const flagModal = (
    <button className="quill-button fun primary outlined" id="edit-flag-button" onClick={toggleFlagModal} type="submit">
      {flag ? flag.label : ''}
    </button>
  );

  const renderActivityForm = () => {
    return(
      <Modal>
        <ActivityForm activity={activity} closeModal={toggleEditActivityModal} submitActivity={submitActivity} />
      </Modal>
    );
  }

  const renderFlagEditModal = () => {
    return(
      <Modal>
        <div className="edit-flag-container">
          <div className="close-button-container">
            <button className="quill-button fun primary contained" id="flag-close-button" onClick={toggleFlagModal} type="submit">x</button>
          </div>
          <DropdownInput
            className="flag-dropdown"
            handleChange={handleFlagChange}
            isSearchable={true}
            label="Development Stage"
            options={flagOptions}
            value={flag}
          />
          <div className="submit-button-container">
            <button className="quill-button fun primary contained" id="flag-submit-button" onClick={handleUpdateFlag} type="submit">
              Submit
            </button>
            <button className="quill-button fun primary contained" id="flag-cancel-button" onClick={toggleFlagModal} type="submit">
              Cancel
            </button>
          </div>
        </div>
      </Modal>
    )
  }

  const generalSettingsRows = (activity) => {
    // format for DataTable to display labels on left side and values on right
    const { passages, prompts, title } = activity
    const fields = [
      { 
        label: 'Title',
        value: title 
      },
      {
        label: 'Development Stage',
        value: flagModal
      },
      {
        label: 'Passage Length',
        value: passages ? `${passages[0].split(' ').length} words` : null
      },
      {
        label: "Because",
        value: prompts ? prompts[0].text : null
      },
      {
        label: "But",
        value: prompts ? prompts[1].text : null
      },
      {
        label: "So",
        value: prompts ? prompts[2].text : null
      },
    ];
    return fields.map(field => {
      const { label, value } = field
      return {
        field: label,
        value
      }
    });
  }

  if(loading) {
    return(
      <div className="loading-spinner-container">
        <Spinner />
      </div>
    );
  }

  if(error) {
    return(
      <div className="error-container">
        <Error error={`${error}`} />
      </div>
    );
  }

  // The header labels felt redundant so passing empty strings and hiding header display
  const dataTableFields = [
    { name: "", attribute:"field", width: "200px" }, 
    { name: "", attribute:"value", width: "400px" }
  ];

  return(
    <div className="activity-settings-container">
      {showEditActivityModal && renderActivityForm()}
      {showEditFlagModal && renderFlagEditModal()}
      <DataTable
        className="activity-general-settings-table"
        headers={dataTableFields}
        rows={generalSettingsRows(activity)}
      />
      <div className="button-container">
        <button className="quill-button fun primary contained" id="edit-activity-button" onClick={toggleEditActivityModal} type="submit">Configure</button>
      </div>
    </div>
  );
}

export default ActivitySettings;
