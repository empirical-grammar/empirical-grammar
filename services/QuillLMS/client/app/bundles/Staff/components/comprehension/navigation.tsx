import * as React from "react";
import { NavLink, } from 'react-router-dom';
import { queryCache } from 'react-query';

import SubmissionModal from './shared/submissionModal';
import ActivityForm from './configureSettings/activityForm';

import { ActivityInterface } from '../../interfaces/comprehensionInterfaces';
import { createActivity } from '../../utils/comprehension/activityAPIs';
import { Modal } from '../../../Shared/index';
import { blankActivity } from '../../../../constants/comprehension';
import { getCsrfToken } from "../../helpers/comprehension";

const Navigation = ({ location, match }) => {
  const { pathname } = location
  const { params, } = match
  const { activityId, } = params
  const [showCreateActivityModal, setShowCreateActivityModal] = React.useState<boolean>(false);
  const [showSubmissionModal, setShowSubmissionModal] = React.useState<boolean>(false);
  const [error, setError] = React.useState<string>('');

  const csrfToken = getCsrfToken();
  localStorage.setItem('csrfToken', csrfToken);

  const checkOverviewActive = () => {
    if(!location) return false;
    return pathname === '/activities' && !showCreateActivityModal;
  }

  const toggleCreateActivityModal = () => {
    setShowCreateActivityModal(!showCreateActivityModal);
  }

  const toggleSubmissionModal = () => {
    setShowSubmissionModal(!showSubmissionModal);
  }

  const submitActivity = (activity: ActivityInterface) => {
    createActivity(activity).then((response) => {
      const { error } = response;
      if(error) setError(error);
      setShowCreateActivityModal(false);
      setShowSubmissionModal(true);

      // update activities cache to display newly created activity
      queryCache.refetchQueries('activities')
    });
  }

  const renderActivityForm = () => {
    return(
      <Modal>
        <ActivityForm activity={blankActivity} closeModal={toggleCreateActivityModal} submitActivity={submitActivity} />
      </Modal>
    );
  }

  const renderSubmissionModal = () => {
    const message = error ? error : 'Submission successful!';
    return <SubmissionModal close={toggleSubmissionModal} message={message} />;
  }

  let activityEditorAndResults

  if (activityId) {
    activityEditorAndResults = (<React.Fragment>
      <p className="menu-label">
        Activity Editor
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/settings`}>
          Activity Settings
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/turk-sessions`}>
          Collect Turk Responses
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/semantic-labels`}>
          Semantic Labels
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/regex-rules`}>
          RegEx Rules
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/plagiarism-rules`}>
          Plagiarism Rules
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/rules`}>
          View All Rules
        </NavLink>
      </ul>
      <p className="menu-label">
        Activity Results
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/stats`}>
          Activity Stats
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/activity-sessions`}>
          View Sessions
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/rules-analysis`}>
          Rules Analysis
        </NavLink>
      </ul>
    </React.Fragment>)
  }

  return(<React.Fragment>
    <section className="left-side-menu">
      <p className="menu-label">
        Tool Views
      </p>
      <ul className="menu-list">
        <NavLink activeClassName='is-active' isActive={checkOverviewActive} to="/activities">
          View Activities
        </NavLink>
        <button className={`create-activity-button ${showCreateActivityModal ? 'is-active' :''}`} onClick={toggleCreateActivityModal} type="submit">
          Create New Activity
        </button>
        <NavLink activeClassName={!showCreateActivityModal ? 'is-active' :''} to="/universal-rules">
          View Universal Rules
        </NavLink>
      </ul>
      {activityEditorAndResults}
    </section>
    {showCreateActivityModal && renderActivityForm()}
    {showSubmissionModal && renderSubmissionModal()}
  </React.Fragment>)
}

export default Navigation
