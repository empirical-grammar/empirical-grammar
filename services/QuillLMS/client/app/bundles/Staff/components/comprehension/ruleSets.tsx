import * as React from "react";
import { Link } from 'react-router-dom'
import { DataTable, Error, Spinner } from 'quill-component-library/dist/componentLibrary';
import { promptStems } from '../../../../constants/comprehension'
const quillCheckmark = `https://assets.quill.org/images/icons/check-circle-small.svg`;
// const quillCheckmark = `${process.env.QUILL_CDN_URL}/images/icons/check-circle-small.svg`;
const fetchAllRuleSetsAPI = 'https://comprehension-dummy-data.s3.us-east-2.amazonaws.com/activities/regex_index.json';

const RuleSets = (props) => {
  const { match } = props;
  const { params } = match;
  const { activityId } = params;
  const [ruleSets, setRuleSets] = React.useState([]);
  const [loading, setLoading] = React.useState(false);
  const [error, setError] = React.useState(null);
  
  const fetchData = async () => {
    try {
      setLoading(true);
      const response = await fetch(fetchAllRuleSetsAPI);
      var json = await response.json();
    } catch (error) {
      setError(error);
      setLoading(false);
    }
    const { regex_entries } = json
    setRuleSets(regex_entries);
    setLoading(false);
  };

  React.useEffect(() => {
    fetchData();
  }, []);

  const getPromptsIcons = (prompts) => {
    const icons = {};
    prompts.forEach(prompt => {
      const { conjunction } = prompt;
      if(promptStems.includes(conjunction)) {
        icons[conjunction] = (<img alt="quill-circle-checkmark" src={quillCheckmark} />)
      } else {
        icons[conjunction] = (<div />);
      }
    });
    return icons;
  }

  const formattedRows = ruleSets.map((rule, i) => {
    const { description, id, prompts, regex_script } = rule;
    const activityLink = (<Link to={`/activities/${activityId}/rulesets/${id}`}>{description}</Link>);
    const promptsIcons = getPromptsIcons(prompts);
    return {
      description: activityLink,
      because_prompt: promptsIcons['because'],
      but_prompt: promptsIcons['but'],
      so_prompt: promptsIcons['so'],
      rule_count: regex_script.length
    }
  });

  if(loading) {
    return(
      <div className="loading-spinner-container">
        <Spinner />
      </div>
    );
  }

  if(error) {
    return(
      <div className="error-container">
        <Error error={`${error}`} />
      </div>
    );
  }

  const dataTableFields = [
    { name: "Description", attribute:"description", width: "400px" }, 
    { name: "Because", attribute:"because_prompt", width: "100px" },
    { name: "But", attribute:"but_prompt", width: "100px" },
    { name: "So", attribute:"so_prompt", width: "100px" },
    { name: "Regex Count", attribute:"rule_count", width: "80px" },
  ];

  return(
    <div className="rulesets-container">
      <DataTable
        className="rulesets-table"
        defaultSortAttribute="description"
        headers={dataTableFields}
        rows={formattedRows}
      />
      <div className="button-container">
        <button className="quill-button fun primary contained" id="add-rule-button" type="submit">Add Rule</button>
      </div>
    </div>
  );
}

export default RuleSets
