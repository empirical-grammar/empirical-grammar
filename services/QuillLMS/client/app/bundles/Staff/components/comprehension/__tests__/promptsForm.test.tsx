import * as React from 'react';
import { shallow } from 'enzyme';

import PromptsForm from '../configureSettings/promptsForm';
import { Input, TextEditor } from '../../../../Shared/index';
import { BUT, BECAUSE, SO, FIRST, SECOND }  from '../../../../../constants/comprehension';

const FEEDBACK = 'At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul.'

const mockProps = {
  activityBecausePrompt: {
    conjunction: BECAUSE,
    text: '1',
    max_attempts: 5,
    max_attempts_feedback: FEEDBACK,
    plagiarism_text: BECAUSE,
    plagiarism_first_feedback: FEEDBACK,
    plagiarism_second_feedback: FEEDBACK
  },
  activityButPrompt: {
    conjunction: BUT,
    text: '2',
    max_attempts: 5,
    max_attempts_feedback: FEEDBACK,
    plagiarism_text: BUT,
    plagiarism_first_feedback: FEEDBACK,
    plagiarism_second_feedback: FEEDBACK
  },
  activitySoPrompt: {
    conjunction: SO,
    text: '3',
    max_attempts: 5,
    max_attempts_feedback: FEEDBACK,
    plagiarism_text: SO,
    plagiarism_first_feedback: FEEDBACK,
    plagiarism_second_feedback: FEEDBACK
  },
  errors: {},
  handleSetPlagiarismFeedback: jest.fn(),
  handleSetPlagiarismText: jest.fn(),
  handleSetPrompt: jest.fn()
};

describe('PromptsForm component', () => {
  const container = shallow(<PromptsForm {...mockProps} />);

  it('should render PromptsForm', () => {
    expect(container).toMatchSnapshot();
  });
  it('should call handleSetPrompt when user types in input fields for because, but and so stems', () => {
    const e = { target: { value: 'test change' } };
    container.find('Input').get(0).props.handleChange(e)
    expect(mockProps.handleSetPrompt).toHaveBeenLastCalledWith(e, BECAUSE);
    container.find('Input').get(3).props.handleChange(e, BUT)
    expect(mockProps.handleSetPrompt).toHaveBeenLastCalledWith(e, BUT);
    container.find('Input').get(6).props.handleChange(e, SO)
    expect(mockProps.handleSetPrompt).toHaveBeenLastCalledWith(e, SO);
  });
  it('should call handleSetPlagiarismText when user types in input fields for because, but and so plagiarism text', () => {
    const e = { target: { value: 'test change' } };
    container.find('TextEditor').get(0).props.handleTextChange(e)
    expect(mockProps.handleSetPlagiarismText).toHaveBeenLastCalledWith(e, BECAUSE);
    container.find('TextEditor').get(1).props.handleTextChange(e, BUT)
    expect(mockProps.handleSetPlagiarismText).toHaveBeenLastCalledWith(e, BUT);
    container.find('TextEditor').get(2).props.handleTextChange(e, SO)
    expect(mockProps.handleSetPlagiarismText).toHaveBeenLastCalledWith(e, SO);
  });
  it('should call handleSetPlagiarismFeedback when user types in input fields for because, but and so plagiarism feedback', () => {
    const e = { target: { value: 'test change' } };
    container.find('Input').get(1).props.handleChange(e)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, FIRST, BECAUSE);
    container.find('Input').get(2).props.handleChange(e, BUT)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, SECOND, BECAUSE);
    container.find('Input').get(4).props.handleChange(e)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, FIRST, BUT);
    container.find('Input').get(5).props.handleChange(e, BUT)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, SECOND, BUT);
    container.find('Input').get(7).props.handleChange(e)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, FIRST, SO);
    container.find('Input').get(8).props.handleChange(e, BUT)
    expect(mockProps.handleSetPlagiarismFeedback).toHaveBeenLastCalledWith(e, SECOND, SO);
  });
});
