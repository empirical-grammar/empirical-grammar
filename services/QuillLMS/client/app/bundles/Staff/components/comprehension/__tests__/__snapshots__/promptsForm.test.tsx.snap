// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptsForm component should render PromptsForm 1`] = `
<Fragment>
  <section
    className="prompt-section"
  >
    <Input
      autoComplete="on"
      className="because-input"
      handleChange={[Function]}
      label="Because Stem"
      value="1"
    />
    <section
      className="plagiarism-section"
    >
      <p
        className="text-editor-label has-text"
      >
        Plagiarism Text
      </p>
      <TextEditor
        ContentState={[Function]}
        EditorState={[Function]}
        handleTextChange={[Function]}
        key="max-attempt-feedback"
        text="because"
      />
      <Input
        autoComplete="on"
        className="primary-feedback-input"
        handleChange={[Function]}
        label="Primary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
      <Input
        autoComplete="on"
        className="secondary-feedback-input"
        handleChange={[Function]}
        label="Secondary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
    </section>
  </section>
  <section
    className="prompt-section"
  >
    <Input
      autoComplete="on"
      className="but-input"
      handleChange={[Function]}
      label="But Stem"
      value="2"
    />
    <section
      className="plagiarism-section"
    >
      <p
        className="text-editor-label has-text"
      >
        Plagiarism Text
      </p>
      <TextEditor
        ContentState={[Function]}
        EditorState={[Function]}
        handleTextChange={[Function]}
        key="max-attempt-feedback"
        text="but"
      />
      <Input
        autoComplete="on"
        className="primary-feedback-input"
        handleChange={[Function]}
        label="Primary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
      <Input
        autoComplete="on"
        className="secondary-feedback-input"
        handleChange={[Function]}
        label="Secondary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
    </section>
  </section>
  <section
    className="prompt-section"
  >
    <Input
      autoComplete="on"
      className="so-input"
      handleChange={[Function]}
      label="So Stem"
      value="3"
    />
    <section
      className="plagiarism-section"
    >
      <p
        className="text-editor-label has-text"
      >
        Plagiarism Text
      </p>
      <TextEditor
        ContentState={[Function]}
        EditorState={[Function]}
        handleTextChange={[Function]}
        key="max-attempt-feedback"
        text="so"
      />
      <Input
        autoComplete="on"
        className="primary-feedback-input"
        handleChange={[Function]}
        label="Primary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
      <Input
        autoComplete="on"
        className="secondary-feedback-input"
        handleChange={[Function]}
        label="Secondary Feedback"
        value="At no point in your rambling, incoherent response were you even close to anything that could be considered a rational thought. I award you no points, and may God have mercy on your soul."
      />
    </section>
  </section>
</Fragment>
`;
